File: leetcode/1358/c++/solution.cpp
/*
    1358. Number of Substrings Containing All Three Characters
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    bool get_bit(uint8_t bitset, size_t position) {
        return (bitset >> position) & 1;
    }

    void set_bit(uint8_t &bitset, size_t position, bool value) {
        bitset = value ? bitset | (1 << position) : bitset & (~(1 << position));
    }

    int number_of_abc_substrings(
        string s, std::unordered_map<std::string, size_t> &abc_sub_str_map) {
        uint8_t marks_abc = 0;
        size_t count = 0;
        for (size_t i = 0; i < s.size() - 2; ++i) {
            for (size_t j = i, k = 1; j < s.size(); ++j) {
                set_bit(marks_abc, static_cast<size_t>(s[j] - 'a'), true);
                if (marks_abc == 7) {
                    count += s.size() - j;
                    break;
                }
            }
            marks_abc = 0;
        }
        return count;
    }

    int numberOfSubstrings(string s) {
        std::unordered_map<std::string, size_t> abc_sub_str_map;
        return number_of_abc_substrings(s, abc_sub_str_map);
    }
};

File: leetcode/1358/rust/solution.rs
/*
    1358. Number of Substrings Containing All Three Characters
    Author: Miguel Angel Bermeo Ayerbe
*/
use std::cmp;

impl Solution {
    pub fn number_of_substrings(s: String) -> i32 {
        let mut marks: Vec<i32> = vec![-1,-1,-1];
        let mut count: i32 = 0;
        let mut i: i32 = 0;
        
        for c_i in s.as_bytes() {
            marks[(c_i - b'a') as usize] = i as i32;
            count += 1+cmp::min(cmp::min(marks[0], marks[1]), marks[2]);
            i+=1;
        }
        count
    }
}